File: projects/group/legacy-flutter/app/lib/agents/user_settings.dart
import 'package:meta/meta.dart';
import 'package:riverpod/riverpod.dart';

@immutable
class TrustedContact {
  const TrustedContact({
    required this.name,
    required this.phone,
    this.relation,
  });

  final String name;
  final String phone;
  final String? relation;

  TrustedContact copyWith({String? name, String? phone, String? relation}) {
    return TrustedContact(
      name: name ?? this.name,
      phone: phone ?? this.phone,
      relation: relation ?? this.relation,
    );
  }
}

@immutable
class UserSettings {
  const UserSettings({
    required this.accountHolder,
    this.overridePin,
    this.emergency,
    this.trusted,
  });

  final String accountHolder;

  /// 4-digit PIN required to override a Guardian block. Set by the
  /// trusted contact, NOT by the elderly user, so they can't be
  /// pressured into entering it on a scam call.
  final String? overridePin;

  final TrustedContact? emergency;
  final TrustedContact? trusted;

  UserSettings copyWith({
    String? accountHolder,
    String? overridePin,
    bool clearOverridePin = false,
    TrustedContact? emergency,
    bool clearEmergency = false,
    TrustedContact? trusted,
    bool clearTrusted = false,
  }) {
    return UserSettings(
      accountHolder: accountHolder ?? this.accountHolder,
      overridePin:
          clearOverridePin ? null : (overridePin ?? this.overridePin),
      emergency: clearEmergency ? null : (emergency ?? this.emergency),
      trusted: clearTrusted ? null : (trusted ?? this.trusted),
    );
  }
}

/// Seeded so the demo's intervention copy ("call your son David") works
/// out of the box. User can edit/overwrite from the Settings screen.
class UserSettingsNotifier extends Notifier<UserSettings> {
  @override
  UserSettings build() {
    return const UserSettings(
      accountHolder: 'Mrs. Wong',
      emergency: TrustedContact(
        name: 'David Wong',
        phone: '+852 9234 5678',
        relation: 'Son',
      ),
      trusted: TrustedContact(
        name: 'Emily Chan',
        phone: '+852 9111 2222',
        relation: 'Daughter',
      ),
    );
  }

  void setEmergency(TrustedContact c) {
    state = state.copyWith(emergency: c);
  }

  void clearEmergency() {
    state = state.copyWith(clearEmergency: true);
  }

  void setTrusted(TrustedContact c) {
    state = state.copyWith(trusted: c);
  }

  void clearTrusted() {
    state = state.copyWith(clearTrusted: true);
  }

  void setOverridePin(String pin) {
    state = state.copyWith(overridePin: pin);
  }

  void clearOverridePin() {
    state = state.copyWith(clearOverridePin: true);
  }

  void setAccountHolder(String name) {
    state = state.copyWith(accountHolder: name);
  }
}

final userSettingsProvider =
    NotifierProvider<UserSettingsNotifier, UserSettings>(
        UserSettingsNotifier.new);

File: projects/group/streamlit_app/guardian/agents/user_settings.py
"""User / account-holder preferences, trusted contacts, override PIN.

1:1 port of ``app/lib/agents/user_settings.dart``.
"""

from __future__ import annotations

from dataclasses import dataclass, field, replace


@dataclass(frozen=True)
class TrustedContact:
    name: str
    phone: str
    relation: str | None = None


@dataclass(frozen=True)
class UserSettings:
    account_holder: str = "Mrs. Wong"
    override_pin: str | None = None
    emergency: TrustedContact | None = None
    trusted: TrustedContact | None = None


class UserSettingsStore:
    """Session-state wrapper around a :class:`UserSettings` value."""

    def __init__(self, initial: UserSettings | None = None) -> None:
        self._state: UserSettings = initial or _default_settings()

    @property
    def state(self) -> UserSettings:
        return self._state

    def set_account_holder(self, name: str) -> None:
        self._state = replace(self._state, account_holder=name)

    def set_emergency(self, contact: TrustedContact) -> None:
        self._state = replace(self._state, emergency=contact)

    def clear_emergency(self) -> None:
        self._state = replace(self._state, emergency=None)

    def set_trusted(self, contact: TrustedContact) -> None:
        self._state = replace(self._state, trusted=contact)

    def clear_trusted(self) -> None:
        self._state = replace(self._state, trusted=None)

    def set_override_pin(self, pin: str) -> None:
        self._state = replace(self._state, override_pin=pin)

    def clear_override_pin(self) -> None:
        self._state = replace(self._state, override_pin=None)


def _default_settings() -> UserSettings:
    return UserSettings(
        account_holder="Mrs. Wong",
        emergency=TrustedContact(
            name="David Wong", phone="+852 9234 5678", relation="Son"
        ),
        trusted=TrustedContact(
            name="Emily Chan", phone="+852 9111 2222", relation="Daughter"
        ),
    )


def default_user_settings() -> UserSettingsStore:
    return UserSettingsStore(_default_settings())
